addons/l10n_eg_edi_eta/models/account_edi_format.py

```diff
@@ -0,0 +1,391 @@
+        access = '%s:%s' % (user, secret)
+        user_and_pass = b64encode(bytes(access, encoding='utf8')).decode('ascii')
+        token_domain = self._l10n_eg_get_eta_token_domain()
+        request_url = '%s/connect/token' % token_domain
+        request_payload = {
+            'grant_type': 'client_credentials',
+        }
+        headers = {'Authorization': f'Basic {user_and_pass}'}
+        try:
+            request_response = requests.post(
+                request_url, data=request_payload, headers=headers, timeout=(5, 10))
+            response_data = request_response.json()
+        except (ValueError, requests.exceptions.ConnectionError, requests.exceptions.MissingSchema, requests.exceptions.Timeout, requests.exceptions.HTTPError) as ex:
+            return {
+                'error': _('Please try again later. Error: \n%s', ex),
+                'blocking_level': 'warning'
+            }
+        if request_response.status_code == 400:
+            return {
+                'error': response_data.get('error', _('Unknown error')),
+                'blocking_level': 'error'
+            }
+        return response_data.get('access_token') or {
+            'error': _('Unable to retrieve ETA token'),
+            'blocking_level': 'warning'
+        }
+
+    def _l10n_eg_validate_different_vats(self, invoice):
+        return invoice.journal_id.l10n_eg_branch_id.vat != invoice.partner_id.vat
+
+    def _l10n_eg_validate_tax_codes(self, invoice):
+        return all(tax.l10n_eg_eta_code for tax in invoice.invoice_line_ids.filtered(lambda x: x.display_type not in ('line_section', 'line_note')).tax_ids)
+
+    def _l10n_eg_validate_uom_codes(self, invoice):
+        return all(account_move_line.product_uom_id.l10n_eg_unit_code_id.code for account_move_line in invoice.invoice_line_ids.filtered(lambda x: x.display_type not in ('line_section', 'line_note')))
+
+    def _l10n_eg_validate_item_codes(self, invoice):
+        return all(account_move_line.product_id.barcode for account_move_line in invoice.invoice_line_ids.filtered(lambda x: x.display_type not in ('line_section', 'line_note')))
+
+    def _l10n_eg_validate_info_address(self, partner_id, issuer=False, invoice_amount=False):
+        fields = ["country_id",
+                  "state_id", "city", "street",
+                  "l10n_eg_building_no"]
+        if (invoice_amount >= self.env.company.l10n_eg_invoicing_threshhold or self._l10n_eg_get_partner_tax_type(partner_id, issuer) != 'P'):
+            fields.append('vat')
+        if issuer:
+            fields.append('l10n_eg_branch_identifier')
+        return all(partner_id[field] for field in fields)
+
+    def _l10n_eg_get_eta_invoice_pdf(self, uuid, invoice):
+        api_domain = self._l10n_eg_get_eta_api_domain()
+        request_url = '%s/api/v1.0/documents/%s/pdf' % (api_domain, url_quote(uuid))
+        token = self._l10n_eg_get_einvoice_token(invoice)
+        if isinstance(token, dict) and token.get('error'):
```

Comment:
try to avoid isinstance a bit.  You could always return a dict and see which key is in there.